### DISCUSSION TITLE
Ego templating library

### DISCUSSION THREAD

**Discussion by menzhessarov** [benbjohnson/wtf#29] [Q&A] [answered]
Hi again! I hope this is my last question so that I won't bother you anymore.

I do not understand what is the purpose of **ego** templating library in the context of the app. What is the advantage that it provides? Is there a reason for not choosing an implementation w/o **ego**?

**Comment by benbjohnson** [comment, marked as answer]
hey again! Great question. There are lots of templating implementations and Go even includes one in its standard library ([html/template](https://golang.org/pkg/html/template/)). There are several things I dislike about most templating libraries (including Go's):

1. They are evaluated at runtime so you don't get the benefit of type checking.
2. They have to be embedded into your application. This is easier with Go 1.16 since it includes file embedding.
3. It's a whole new language to learn on top of Go.

The [ego](https://github.com/benbjohnson/ego) templating library is modeled after the [ERb](https://docs.ruby-lang.org/en/2.3.0/ERB.html) template library from Ruby and it's a really thin layer over the Go language. It transpiles `.ego` files to `.go` files at compile time so that you end up with templates that are pure Go. That makes them extremely fast, automatically embedded, and provides all of Go's type checking capabilities (since they're just Go files).

There's even a nifty trick where `ego` embeds `//go:line` commands to show correct line numbers in Go errors.

**Reply by menzhessarov** [reply]
Thank you for your time. This is all super insightful to me. Related question to `html` package in the app.

```
prev := current - 1
if prev <= 0 {
  prev = 1
}
 
next := current + 1
if next >= pageN {
  next = pageN
}
```
In `Pagination`'s `Render` method you do not use `prev` or `next` after this code. Is it intentional?

**Reply by benbjohnson** [reply]
Good catch. That's just a mistake on my end. 👍
